File: WMF Framework/WMFOnThisDayEventsFetcher.m
#import "WMFOnThisDayEventsFetcher.h"
#import "WMFFeedOnThisDayEvent.h"
#import <WMF/WMF-Swift.h>
#import <WMF/WMFLegacySerializer.h>

@implementation WMFOnThisDayEventsFetcher

+ (NSSet<NSString *> *)supportedLanguages {
    static dispatch_once_t onceToken;
    static NSSet<NSString *> *supportedLanguages;
    dispatch_once(&onceToken, ^{
        supportedLanguages = [NSSet setWithObjects:@"en", @"de", @"sv", @"fr", @"es", @"ru", @"pt", @"ar", @"uk", @"tr", nil];
    });
    return supportedLanguages;
}

+ (BOOL)isOnThisDaySupportedByLanguage:(NSString *)languageCode {
    return [[WMFOnThisDayEventsFetcher supportedLanguages] containsObject:languageCode];
}

- (void)fetchOnThisDayEventsForURL:(NSURL *)siteURL month:(NSUInteger)month day:(NSUInteger)day failure:(WMFErrorHandler)failure success:(void (^)(NSArray<WMFFeedOnThisDayEvent *> *announcements))success {
    NSParameterAssert(siteURL);
    if (siteURL == nil || siteURL.wmf_languageCode == nil || ![WMFOnThisDayEventsFetcher isOnThisDaySupportedByLanguage:siteURL.wmf_languageCode] || month < 1 || day < 1) {
        NSError *error = [WMFFetcher invalidParametersError];
        failure(error);
        return;
    }

    NSString *monthString = [NSString stringWithFormat:@"%lu", (unsigned long)month];
    NSString *dayString = [NSString stringWithFormat:@"%lu", (unsigned long)day];
    NSArray<NSString *> *path = @[@"feed", @"onthisday", @"events", monthString, dayString];
    NSURL *url = [self.configuration feedContentAPIURLForURL:siteURL appendingPathComponents:path];
    [self.session getJSONDictionaryFromURL:url ignoreCache:YES completionHandler:^(NSDictionary<NSString *,id> * _Nullable result, NSHTTPURLResponse * _Nullable response, NSError * _Nullable error) {
        if (error) {
            failure(error);
            return;
        }
        
        if (response.statusCode == 304) {
            failure([WMFFetcher noNewDataError]);
            return;
        }
        
        NSError *serializerError = nil;
        NSArray *events = [WMFLegacySerializer modelsOfClass:[WMFFeedOnThisDayEvent class] fromArrayForKeyPath:@"events" inJSONDictionary:result languageVariantCode: url.wmf_languageVariantCode error:&serializerError];
        if (serializerError) {
            failure(serializerError);
            return;
        }
        
        success(events);
    }];
}

@end

File: WMF Framework/WMFOnThisDayEventsFetcher.swift
import Foundation
import WMFData

/// Fetches the on-this-day events for the Explore feed.
///
/// The fetcher uses `WMFOnThisDayDataController` and converts the result into `WMFFeedOnThisDayEvent` objects.
/// Remove this class when the Explore feed is removed.
@objc(WMFOnThisDayEventsFetcher)
public final class WMFOnThisDayEventsFetcher: NSObject {

    /// The languages with an on-this-day feed.
    private static let supportedLanguages: Set<String> = ["en", "de", "sv", "fr", "es", "ru", "pt", "ar", "uk", "tr"]

    private let dataController: WMFOnThisDayDataController

    @objc public override init() {
        dataController = WMFOnThisDayDataController.shared
        super.init()
    }

    public init(dataController: WMFOnThisDayDataController) {
        self.dataController = dataController
        super.init()
    }

    @objc(isOnThisDaySupportedByLanguage:)
    public static func isOnThisDaySupported(by languageCode: String) -> Bool {
        return supportedLanguages.contains(languageCode)
    }

    /// Fetch the events for a month and a day.
    @objc(fetchOnThisDayEventsForURL:month:day:failure:success:)
    public func fetchOnThisDayEvents(forURL siteURL: URL, month: UInt, day: UInt, failure: @escaping (Error) -> Void, success: @escaping ([WMFFeedOnThisDayEvent]) -> Void) {
        guard let languageCode = siteURL.wmf_languageCode,
              WMFOnThisDayEventsFetcher.isOnThisDaySupported(by: languageCode),
              month >= 1, day >= 1 else {
            failure(RequestError.invalidParameters)
            return
        }

        let languageVariantCode = siteURL.wmf_languageVariantCode
        let project = WMFProject.wikipedia(WMFLanguage(languageCode: languageCode, languageVariantCode: languageVariantCode))
        let dataController = self.dataController

        Task {
            do {
                let response = try await dataController.fetchOnThisDay(project: project, month: Int(month), day: Int(day))
                success(response.events.map { WMFFeedOnThisDayEvent(event: $0, languageVariantCode: languageVariantCode) })
            } catch {
                failure(error)
            }
        }
    }
}
